dask_cuda/explicit_comms/dataframe_shuffle.py

```diff
@@ -154,24 +197,139 @@ def dataframe_shuffle(df: DataFrame, column_names: List[str]) -> DataFrame:
+    output_keys = []
+    for input_key in df.__dask_keys__():
+        output_key = (name, input_key[1])
+        dsk[output_key] = (
+            shuffle_group,
+            input_key,
+            column_names,
+            0,
+            npartitions,
+            npartitions,
+            ignore_index,
+            npartitions,
+        )
+        output_keys.append(output_key)
+    df_groups = compute_as_if_collection(type(df), dsk, output_keys, sync=False)
+    wait(df_groups)
+    for f in df_groups:  # Check for errors
+        if f.status == "error":
+            f.result()  # raise exception
+
+    # Step (b): find out which workers has what part of `df_groups`,
+    #           find the number of output each worker should have,
+    #           and submit `local_shuffle()` on each worker.
+    key_to_part = dict([(str(part.key), part) for part in df_groups])
+    in_parts = defaultdict(list)  # Map worker -> [list of futures]
+    for key, workers in c.client.who_has(df_groups).items():
+        # Note, if multiple workers have the part, we pick the first worker
+        in_parts[first(workers)].append(key_to_part[key])
+
+    # Let's create a dict that specifices the number of partitions each worker has
+    in_nparts = {}
+    workers = set()  # All ranks that have a partition of `df`
+    for rank, worker in enumerate(c.worker_addresses):
+        nparts = len(in_parts.get(worker, ()))
+        if nparts > 0:
+            in_nparts[rank] = nparts
+            workers.add(rank)
+
+    # Find the output partitions for each worker
+    div = npartitions // len(workers)
+    rank_to_out_part_ids = {}  # rank -> [list of partition id]
+    for i, rank in enumerate(workers):
+        rank_to_out_part_ids[rank] = list(range(div * i, div * (i + 1)))
+    for rank, i in zip(workers, range(div * len(workers), npartitions)):
+        rank_to_out_part_ids[rank].append(i)
+
+    # Run `local_shuffle()` on each worker
+    result_futures = {}
+    for rank, worker in enumerate(c.worker_addresses):
+        if rank in workers:
+            result_futures[rank] = c.submit(
+                worker,
+                local_shuffle,
+                workers,
+                in_nparts,
+                in_parts[worker],
+                rank_to_out_part_ids,
+                ignore_index,
+            )
+    distributed.wait(list(result_futures.values()))
+
+    # Step (c): extract individual dataframe-partitions
+    ret = []
+    for rank, parts in rank_to_out_part_ids.items():
+        for i in range(len(parts)):
+            ret.append(delayed(getitem)(result_futures[rank], i))
+    del result_futures
+    return dask.dataframe.from_delayed(ret, verify_meta=False).persist()
+
+
+def rearrange_by_column_tasks_wrapper(
+    df, column, max_branch=32, npartitions=None, ignore_index=False
+):
+    """Function wrapper that dispatch the shuffle to explicit-comms.
+
+    Notice, this is monkey patched into Dask at dask_cuda import
```

Comment:
Again here, I think we need to consider whether this is safe to do, as per my comment in `__init__.py`.

Reply:
After my resent commit, I think this is safe -- the wrapper function simply pass-through `*args, **kwargs` when `DASK_EXPLICIT_COMMS=False`.

Reply:
Thanks Mads!